packages/beacon-node/src/chain/validation/executionPayloadEnvelope.ts

```diff
@@ -0,0 +1,120 @@
+import {
+  computeStartSlotAtEpoch,
+  createSingleSignatureSetFromComponents,
+  getExecutionPayloadSigningRoot,
+} from "@lodestar/state-transition";
+import {gloas} from "@lodestar/types";
+import {toRootHex} from "@lodestar/utils";
+import {ExecutionPayloadEnvelopeError, ExecutionPayloadEnvelopeErrorCode, GossipAction} from "../errors/index.js";
+import {IBeaconChain} from "../index.js";
+
+export async function validateApiExecutionPayloadEnvelope(
+  chain: IBeaconChain,
+  executionPayloadEnvelope: gloas.SignedExecutionPayloadEnvelope
+): Promise<void> {
+  return validateExecutionPayloadEnvelope(chain, executionPayloadEnvelope);
+}
+
+export async function validateGossipExecutionPayloadEnvelope(
+  chain: IBeaconChain,
+  executionPayloadEnvelope: gloas.SignedExecutionPayloadEnvelope
+): Promise<void> {
+  return validateExecutionPayloadEnvelope(chain, executionPayloadEnvelope);
+}
+
+async function validateExecutionPayloadEnvelope(
+  chain: IBeaconChain,
+  executionPayloadEnvelope: gloas.SignedExecutionPayloadEnvelope
+): Promise<void> {
+  const envelope = executionPayloadEnvelope.message;
+  const {builderIndex, payload} = envelope;
+  const blockRootHex = toRootHex(envelope.beaconBlockRoot);
+
+  //  GLOAS: [IGNORE] The envelope's block root `envelope.block_root` has been seen (via
+  //  gossip or non-gossip sources) (a client MAY queue payload for processing once
+  //  the block is retrieved).
+  // TODO GLOAS: Need to review this
+  const block = chain.forkChoice.getBlock(envelope.beaconBlockRoot);
+  if (block === null) {
+    throw new ExecutionPayloadEnvelopeError(GossipAction.IGNORE, {
+      code: ExecutionPayloadEnvelopeErrorCode.BLOCK_ROOT_UNKNOWN,
+      blockRoot: blockRootHex,
+    });
+  }
+
+  //  [IGNORE] The node has not seen another valid
+  //  `SignedExecutionPayloadEnvelope` for this block root from this builder.
+  if (chain.seenExecutionPayloadEnvelopes.isKnown(blockRootHex, builderIndex)) {
```

Comment:
there is no reason to have builder index here, we add the payload to seen cache after signature is verified and we check beforehand that it's same as `bid.builder_index` which means this can only ever be one builder


```suggestion
  if (chain.seenExecutionPayloadEnvelopes.isKnown(blockRootHex)) {
```